rust/worker/src/blockstore/arrow/block/types.rs

```diff
@@ -203,108 +240,431 @@ impl Block {
+        let options =
+            match arrow::ipc::writer::IpcWriteOptions::try_new(64, false, MetadataVersion::V5) {
+                Ok(options) => options,
+                Err(e) => {
+                    return Err(BlockSaveError::ArrowError(e));
+                }
+            };
+
+        let writer = arrow::ipc::writer::FileWriter::try_new_with_options(
+            &mut writer,
+            &self.data.schema(),
+            options,
+        );
         let mut writer = match writer {
             Ok(writer) => writer,
             Err(e) => {
-                // TODO: Return a proper error
-                panic!("Error creating writer: {:?}", e)
+                return Err(BlockSaveError::ArrowError(e));
             }
         };
         match writer.write(&self.data) {
             Ok(_) => match writer.finish() {
                 Ok(_) => return Ok(()),
                 Err(e) => {
-                    panic!("Error finishing writer: {:?}", e);
+                    return Err(BlockSaveError::ArrowError(e));
                 }
             },
             Err(e) => {
-                panic!("Error writing data: {:?}", e);
+                return Err(BlockSaveError::ArrowError(e));
             }
         }
     }
 
-    pub fn to_bytes(&self) -> Vec<u8> {
+    /// Convert the block to bytes in Arrow IPC format
+    pub fn to_bytes(&self) -> Result<Vec<u8>, BlockToBytesError> {
         let mut bytes = Vec::new();
         // Scope the writer so that it is dropped before we return the bytes
         {
             let mut writer =
-                arrow::ipc::writer::FileWriter::try_new(&mut bytes, &self.data.schema())
-                    .expect("Error creating writer");
-            writer.write(&self.data).expect("Error writing data");
-            writer.finish().expect("Error finishing writer");
+                match arrow::ipc::writer::FileWriter::try_new(&mut bytes, &self.data.schema()) {
+                    Ok(writer) => writer,
+                    Err(e) => {
+                        return Err(BlockToBytesError::ArrowError(e));
+                    }
+                };
+            match writer.write(&self.data) {
+                Ok(_) => {}
+                Err(e) => {
+                    return Err(BlockToBytesError::ArrowError(e));
+                }
+            }
+            match writer.finish() {
+                Ok(_) => {}
+                Err(e) => {
+                    return Err(BlockToBytesError::ArrowError(e));
+                }
+            }
         }
-        bytes
+        Ok(bytes)
+    }
+
+    pub fn from_bytes(bytes: &[u8], id: Uuid) -> Result<Self, BlockLoadError> {
+        return Self::from_bytes_internal(bytes, id, false);
     }
 
-    pub fn from_bytes(bytes: &[u8], id: Uuid) -> Result<Self, Box<dyn ChromaError>> {
+    pub fn from_bytes_with_validation(bytes: &[u8], id: Uuid) -> Result<Self, BlockLoadError> {
```

Comment:
the _with_validation methods for both from_bytes and load_ are meant to be in testing only